pubsub/README.md

```diff
@@ -149,22 +156,171 @@ economics (see e.g.
+> within the timecache interval, as they will get treated as duplicates. This
+> consequence may or may not be relevant to the application at hand.
+> Reference: [#116](https://github.com/libp2p/specs/issues/116).
+
+**Note that the availability of these fields on the `Message` object will depend
+on the [signature policy](#signature-policy) configured for the topic.**
+
+Whichever the choice, it is crucial that **all peers** participating in a topic
+implement identical message ID calculation logic, or the topic will malfunction.
+
+> **[[ Implementation note ]]:** At the time of writing this section,
+> go-libp2p-pubsub (reference implementation of this spec) only allows
+> configuring a single top-level `message_id_fn`. This function may, however,
+> vary its behaviour based on the topic (contained inside its `Message`)
+> argument. Thus, it's feasible to implement a per-topic policy using branch
+> selection control flow logic. In the near future, go-libp2p-pubsub plans to
+> push down the configuration of the `message_id_fn` to the topic level. Other
+> implementations are encouraged to do the same.
+
 ## Message Signing
 
-Messages can be optionally signed, and it is up to the peer whether to accept and forward
-unsigned messages.
+Signature behavior is configured in two axes: signature creation, and signature
+verification.
+
+**Signature creation.** There are two configurations possible:
+
+* `Sign`: when publishing a message, perform **origin-stamping** and produce a
+  signature.
+* `NoSign`: when publishing a message, do not perform **origin-stamping** and
+  do not produce a signature.
 
 For signing purposes, the `signature` and `key` fields are used:
 - The `signature` field contains the signature.
-- The `key` field contains the signing key when it cannot be inlined in the source peer ID.
-  When present, it must match the peer ID.
+- The `key` field contains the signing key when it cannot be inlined in
+  the source peer ID (`from`). When present, it must match the peer ID.
+
+The signature is computed over the marshalled message protobuf _excluding_ the
+`signature` field itself.
+
+This includes any fields that are not recognized, but still included in the
+marshalled data.
 
-The signature is computed over the marshalled message protobuf _excluding_ the key field.
 The protobuf blob is prefixed by the string `libp2p-pubsub:` before signing.
 
+> **[[ Margin note: ]]** Protobuf serialization is non-deterministic/canonical,
+> and the same data structure may result in different, valid serialised bytes
+> across implementations, as well as other issues. In the near future, the
+> signature creation and verification algorithm will be made deterministic.
+
+**Signature verification.** There are two configurations possible:
+
+* `Strict`: either expect or not expect a signature.
+* `Lax` (legacy, insecure, underterministic, to be deprecated): accept a signed
```

Comment:
It would be nice to export some of this language to a separate section which simply says that all fields that are set must be valid for their type - `seqno` must be 8 bytes, if the signature is set it *must* be valid - these would be pre-conditions to even consider the message for propagation before the "modes" are considered - the "lax" vs "strict" distinction then becomes a requirement for a signature or not.

Reply:
Sure, that would be a welcome improvement. All protobuf types are `bytes` and `string`s, which express no constrains in and of themselves.

Reply:
Mind submitting a PR against master when this lands (i.e. now)?